devito/operations/solve.py

```diff
@@ -0,0 +1,120 @@
+from functools import singledispatch
+
+import sympy
+
+from devito.logger import warning
+from devito.finite_differences.differentiable import Add, Mul, EvalDerivative
+from devito.finite_differences.derivative import Derivative
+from devito.tools import as_tuple
+
+__all__ = ['solve', 'linsolve']
+
+
+class SolveError(Exception):
+    """Base class for exceptions in this module."""
+    pass
+
+
+def solve(eq, target, **kwargs):
+    """
+    Algebraically rearrange an Eq w.r.t. a given symbol.
+
+    This is a wrapper around ``sympy.solve``.
+
+    Parameters
+    ----------
+    eq : expr-like
+        The equation to be rearranged.
+    target : symbol
+        The symbol w.r.t. which the equation is rearranged. May be a `Function`
+        or any other symbolic object.
+    **kwargs
+        Symbolic optimizations applied while rearranging the equation. For more
+        information. refer to ``sympy.solve.__doc__``.
+    """
+
+    if hasattr(eq, 'rhs'):
```

Comment:
the pythonic way is a `try-except(AttributeError)`. We should always avoid `hasattr` and `getattr` for as much as we can